middle_end/flambda2/nominal/name_abstraction.ml

```diff
@@ -14,71 +14,55 @@
 (*                                                                        *)
 (**************************************************************************)
 
-[@@@ocaml.warning "+a-4-30-40-41-42"]
+[@@@ocaml.warning "+a-30-40-41-42"]
 
-module type Term = sig
-  include Contains_names.S
+type ('bindable, 'term) t = 'bindable * 'term
 
-  include Contains_ids.S with type t := t
+let create bindable term = bindable, term
 
-  val print : Format.formatter -> t -> unit
-end
+let[@inline always] pattern_match ~freshen_bindable ~swap_bindable
+    ~apply_renaming_term (bindable, term) ~f =
+  let fresh_bindable = freshen_bindable bindable in
+  let renaming = swap_bindable bindable ~guaranteed_fresh:fresh_bindable in
+  let fresh_term = apply_renaming_term term renaming in
+  f fresh_bindable fresh_term
 
-module Make (Bindable : Bindable.S) (Term : Term) = struct
-  type t = Bindable.t * Term.t
+let[@inline always] pattern_match_pair ~freshen_bindable ~swap_bindable
+    ~apply_renaming_term (bindable0, term0) (bindable1, term1) ~f =
+  let fresh_bindable = freshen_bindable bindable0 in
+  let renaming0 = swap_bindable bindable0 ~guaranteed_fresh:fresh_bindable in
+  let renaming1 = swap_bindable bindable1 ~guaranteed_fresh:fresh_bindable in
+  let fresh_term0 = apply_renaming_term term0 renaming0 in
+  let fresh_term1 = apply_renaming_term term1 renaming1 in
+  f fresh_bindable fresh_term0 fresh_term1
 
-  let create name term = name, term
-
-  let[@inline always] pattern_match (name, term) ~f =
-    let fresh_name = Bindable.rename name in
-    let perm = Bindable.name_permutation name ~guaranteed_fresh:fresh_name in
-    let fresh_term = Term.apply_renaming term perm in
-    f fresh_name fresh_term
-
-  let [@ocamlformat "disable"] print ppf t =
-    pattern_match t ~f:(fun name term ->
-      Format.fprintf ppf "@[<hov 1>%s@<1>%s%s%a%s@<1>%s%s@ %a@]"
+let print ~print_bindable ~print_term ~freshen_bindable ~swap_bindable
```

Comment:
As long as we're looking at this code, I'd like some support for non-freshening printing. It's particularly annoying that adding debug printing changes the output.